packages/git/src/browser/diff/git-diff-contribution.ts

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { CommandRegistry, Command, MenuModelRegistry, SelectionService } from "@theia/core/lib/common";
+import { FrontendApplication, AbstractViewContribution } from '@theia/core/lib/browser';
+import { WidgetManager } from '@theia/core/lib/browser/widget-manager';
+import { injectable, inject } from "inversify";
+import { GitDiffViewOptions } from './git-diff-model';
+import { GitDiffWidget } from './git-diff-widget';
+import { NAVIGATOR_CONTEXT_MENU } from '@theia/navigator/lib/browser/navigator-menu';
+import { UriCommandHandler, FileSystemCommandHandler } from '@theia/workspace/lib/browser/workspace-commands';
+import { GitQuickOpenService } from '../git-quick-open-service';
+
+export namespace GitDiffCommands {
+    export const OPEN_DIFF: Command = {
+        id: 'git-diff:toggle',
+        label: 'Diff'
+    };
+    export const OPEN_FILE_DIFF: Command = {
+        id: 'git-diff:open-file-diff',
+        label: 'Compare with...'
+    };
+}
+
+export const GIT_DIFF = "git-diff";
+
+@injectable()
+export class GitDiffContribution extends AbstractViewContribution<GitDiffWidget> {
+
+    constructor(
+        @inject(SelectionService) protected readonly selectionService: SelectionService,
+        @inject(WidgetManager) protected readonly widgetManager: WidgetManager,
+        @inject(FrontendApplication) protected readonly app: FrontendApplication,
+        @inject(GitQuickOpenService) protected readonly quickOpenService: GitQuickOpenService
+    ) {
+        super({
+            widgetId: GIT_DIFF,
+            widgetName: 'Git diff',
+            defaultWidgetOptions: {
+                area: 'left',
+                rank: 400
+            }
+        });
+    }
+
+    registerMenus(menus: MenuModelRegistry): void {
+        menus.registerMenuAction([...NAVIGATOR_CONTEXT_MENU, '5_history'], {
+            commandId: GitDiffCommands.OPEN_FILE_DIFF.id
+        });
+    }
+
+    registerCommands(commands: CommandRegistry): void {
+        commands.registerCommand(GitDiffCommands.OPEN_FILE_DIFF, this.newFileHandler({
+            execute: async uri => {
+                await this.quickOpenService.chooseTagsAndBranches(
+                    (fromRevision, toRevision) => {
+                        const fileUri = uri.toString();
```

Comment:
What does command should do? I did not find much about the requirements in the documentation. (I might have looked an outdated one though.)

We cannot just treat all files as text files when calculating the diff.

Reply:
It opens a list of tags and branches to compare the current state against.

Reply:
Does that mean we need the content of that file at two different points in time?